lib/websocket/connection.js

```diff
@@ -0,0 +1,551 @@
+ * @param {import('net').Socket|import('tls').TLSSocket|undefined} socket
+ */
+function failWebsocketConnection (controller, socket) {
+  controller.abort()
+
+  if (socket && !socket.destroyed) {
+    socket.destroy()
+  }
+}
+
+/**
+ * @see https://websockets.spec.whatwg.org/#feedback-from-the-protocol
+ * @param {import('./websocket').WebSocket} ws
+ */
+function whenConnectionEstablished (ws) {
+  const { [kResponse]: response } = ws
+
+  // 1. Change the ready state to OPEN (1).
+  ws[kReadyState] = states.OPEN
+
+  // 2. Change the extensions attribute’s value to the extensions in use, if
+  //    it is not the null value.
+  // https://datatracker.ietf.org/doc/html/rfc6455#section-9.1
+  const extensions = response.headersList.get('sec-websocket-extensions')
+
+  if (extensions !== null) {
+    ws[kExtensions] = extensions
+  }
+
+  // 3. Change the protocol attribute’s value to the subprotocol in use, if
+  //    it is not the null value.
+  // https://datatracker.ietf.org/doc/html/rfc6455#section-1.9
+  const protocol = response.headersList.get('sec-websocket-protocol')
+
+  if (protocol !== null) {
+    ws[kProtocol] = protocol
+  }
+
+  // 4. Fire an event named open at the WebSocket object.
+  fireEvent('open', ws)
+
+  if (channels.open.hasSubscribers) {
+    channels.open.publish({
+      address: response.socket.address(),
+      protocol,
+      extensions
+    })
+  }
+}
+
+/**
+ * @see https://websockets.spec.whatwg.org/#feedback-from-the-protocol
+ * @param {import('./websocket').WebSocket} ws
+ */
+function receiveData (ws) {
+  const { [kResponse]: response } = ws
+
+  /** @type {WebsocketFrame|undefined} */
+  let frame
+
+  const partialFrame = { chunks: [], length: 0, payloadLength: undefined }
+
+  response.socket.on('data', (chunk) => {
+    if (!frame) {
+      partialFrame.chunks.push(chunk)
+      partialFrame.length += chunk.length
+
+      if (partialFrame.length < 2) {
+        return // Can't read payload length yet
+      } else if (partialFrame.payloadLength !== undefined) {
+        if (partialFrame.payloadLength === 126) {
+          if (partialFrame.length < 4) {
+            return
+          }
+        } else if (partialFrame.payloadLength === 127) {
+          if (partialFrame.length < 10) {
+            return
+          }
+        }
+      } else {
+        const data = Buffer.concat(partialFrame.chunks, partialFrame.length)
+        const payloadLength = data[1] & 0x7F
+
+        partialFrame.chunks = [data]
+        partialFrame.payloadLength = payloadLength
+
+        if (payloadLength === 126) {
+          if (partialFrame.length < 4) {
+            return
+          }
+        } else if (payloadLength === 127) {
+          if (partialFrame.length < 10) {
+            return
+          }
+        }
+      }
+    }
+
+    let receivedFrame
+
+    if (partialFrame.payloadLength !== undefined) {
+      receivedFrame = WebsocketFrame.from(Buffer.concat(partialFrame.chunks, partialFrame.length))
```

Comment:
What happens if you receive a chunk like `82008200` that is two empty frames in the same data chunk? I did not test it but it does not seem correct.

Reply:
to be honest I'd prefer to fix this in a followup pr. Behavior like this doesn't seem common and there's currently a lot of "weird behavior" things we aren't testing (the wpts are extremely lacking).

Reply:
Your call, but trust me, this is common. Not two empty frames in the same data chunk, that is just an example for easy reproduction, but multiple frames in the same data chunk and parts of them. A working frame parser is a basic thing, otherwise you might miss messages, leak memory and sockets, etc. It's like shipping HTTP with a broken HTTP parser.

Try to subscribe to the websocket feed of a popular cryptocurrency exchange:

```js
const ws = new WebSocket('wss://ws-feed.exchange.coinbase.com');

ws.addEventListener('open', function () {
  ws.send(
    JSON.stringify({
      type: 'subscribe',
      product_ids: ['BTC-USD'],
      channels: ['full']
    })
  );
});

ws.addEventListener('message', function (event) {
  console.log(event.data);
});
```

Reply:
Thanks! We aren't leaking memory/sockets, but messages are being missed. I need to move the implementation to something that supports buffering & reading the body only until the application data length.

Reply:
Okay, this is finally fixed. I took a lot of inspiration from ws (which I am becoming ever more grateful for) so everything is working pretty smoothly. Still needs validation and a lot of work, but much better than the old solution.

This should also fix the issue where sending < 2 bytes would cause an issue.